kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/WorkflowActivityRunner.kt

```diff
@@ -0,0 +1,145 @@
+   * Read it back out in [FragmentActivity.onCreate], to serve as the
+   * final argument to [FragmentActivity.setContentWorkflow].
+   */
+  val asParcelable: Parcelable get() = PickledWorkflow(model.lastSnapshot)
+
+  /**
+   * To be called from [FragmentActivity.onBackPressed], to give the managed
+   * [Workflow] access to back button events.
+   *
+   * e.g.:
+   *
+   *    override fun onBackPressed() {
+   *      if (!workflowViewModel.onBackPressed(this)) super.onBackPressed()
+   *    }
+   */
+  fun onBackPressed(activity: Activity): Boolean {
+    return HandlesBack.Helper.onBackPressed(activity.findViewById(R.id.workflow_layout))
+  }
+}
+
+/**
+ * Call this method from [FragmentActivity.onCreate], instead of [FragmentActivity.setContentView].
+ * It creates a [WorkflowActivityRunner] for this activity, if one doesn't already exist, and
+ * sets a view driven by that model as the content view.
+ *
+ * Hold onto the [WorkflowActivityRunner] returned and:
+ *
+ *  - Call [WorkflowActivityRunner.onBackPressed] from [FragmentActivity.onBackPressed] to allow
+ *    workflows to handle back button events. (See [HandlesBack] for more details.)
+ *
+ *  - Write [WorkflowActivityRunner.asParcelable] to the bundle passed to
+ *    [FragmentActivity.onSaveInstanceState]. You'll read that [parcelable][PickledWorkflow]
+ *    back in [FragmentActivity.onCreate], for use by the next call to this method.
+ *
+ *  e.g.:
+ *
+ *     class MainActivity : AppCompatActivity() {
+ *       private lateinit var runner: WorkflowRunner<*, *>
+ *
+ *       override fun onCreate(savedInstanceState: Bundle?) {
+ *         super.onCreate(savedInstanceState)
+ *
+ *         val restored = savedInstanceState?.getParcelable<PickledWorkflow>("WORKFLOW")
+ *         runner = setContentWorkflow(MyViewRegistry, MyRootWorkflow(), restored)
+ *       }
+ *
+ *       override fun onBackPressed() {
+ *         if (!runner.onBackPressed(this)) super.onBackPressed()
+ *       }
+ *
+ *       override fun onSaveInstanceState(outState: Bundle) {
+ *         super.onSaveInstanceState(outState)
+ *         outState.putParcelable("WORKFLOW", runner.asParcelable)
+ *       }
+ *     }
+ */
+@CheckResult
+fun <InputT : Any, OutputT : Any, RenderingT : Any> FragmentActivity.setContentWorkflow(
+  viewRegistry: ViewRegistry,
+  workflow: Workflow<InputT, OutputT, RenderingT>,
+  initialInput: InputT,
+  restored: PickledWorkflow?
+): WorkflowActivityRunner<OutputT, RenderingT> {
+  val factory = WorkflowViewModel.Factory(viewRegistry, workflow, initialInput, restored)
+
+  // We use an Android lifecycle ViewModel to shield ourselves from configuration changes.
+  // ViewModelProviders.of() uses the factory to instantiate a new instance only
+  // on the first call for this activity, and it stores that instance for repeated use
+  // until this activity is finished.
+
+  @Suppress("UNCHECKED_CAST")
+  val viewModel = ViewModelProviders.of(this, factory)[WorkflowViewModel::class.java]
```

Comment:
I'm surprised to see subscripts here since I didn't think `ViewModelProviders.of` returned a list. Was the IDE complaining about a `.get` call?

Reply:
No list involved here. The complaint is because of the parameter types, we're casting `WorkflowViewModel<*, *>` to `WorkflowViewModel<O, R>`

Reply:
I mean seeing `[…]` looks weird since there's no list.

Reply:
It has a get method, and is essentially intended to act as a map indexed by the class names. Google's sample docs used this consistently, so I figured I'd do the same.

Reply:
Ah, that seems fine then. I hadn't seen that the last time I read them.